File: auth-service/src/delivery/error_handler.ts
import { BadRequestError } from './../error';
import { Response, Request, NextFunction } from 'express';

type errResponse = {
    code: string,
    message: string,
}

export function errorHandler(err: TypeError | BadRequestError | SyntaxError, req: Request, res: Response, next: NextFunction): void {
    let resp: errResponse = {
        code: "-1",
        message: "fatal error! please contact the service owner"
    };
    
    if(err instanceof BadRequestError) {
        res.status(400)
        resp.code = err._statusCode;
        resp.message = err.message;
        res.json(resp)
        return
    }
    
    if(err instanceof SyntaxError) {
        res.status(400)
        resp.code = "400";
        resp.message = err.message;
        res.json(resp)
        return
    }

    console.log("FATAL ERROR:", err) // should be contains all request meta data for tracing & debugging
    res.status(500)
    res.json(resp)
}

export function errorNotFound(req: Request, res: Response, next: NextFunction) {
    res.status(404)
    res.json({
        code: "404",
        message: "not found"
    })
}

File: fetch-service/internal/delivery/error_handler.go
package delivery

import (
	"fmt"
	"net/http"

	"github.com/bagusays/efishery-technical-test/internal"
	"github.com/labstack/echo/v4"
	"github.com/spf13/cast"
)

type ErrResponse struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// ErrorHandler :nodoc:
func ErrorHandler(err error, c echo.Context) {
	if c.Response().Committed || err == nil {
		return
	}

	if c.Response().Size > 0 {
		return
	}

	httpStatusCode := http.StatusInternalServerError
	resp := ErrResponse{
		Code:    "-1",
		Message: "fatal error! please contact the service owner",
	}

	switch e := err.(type) {
	case *internal.Error:
		httpStatusCode = http.StatusBadRequest
		resp.Message = e.Error()
		resp.Code = e.StatusCode()
	case *echo.HTTPError:
		httpStatusCode = e.Code
		resp.Message = e.Error()
		resp.Code = cast.ToString(e.Code)
	default:
		fmt.Println("FATAL ERROR:", e.Error())
	}

	_ = c.JSON(httpStatusCode, resp)
}
